cli/registry.go

```diff
@@ -0,0 +1,247 @@
+		return err
+	}
+
+	ctx := context.Background()
+	docker, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return fmt.Errorf("Couldn't create docker client\n%+w", err)
+	}
+
+	return docker.CopyToContainer(ctx, ID, defaultFullRegistriesPath, bytes.NewReader(d), types.CopyToContainerOptions{
+		AllowOverwriteDirWithFile: true,
+		CopyUIDGID:                false,
+	})
+}
+
+// createRegistry creates a registry, or connect the k3d network to an existing one
+func createRegistry(spec ClusterSpec) (string, error) {
+	netName := k3dNetworkName(spec.ClusterName)
+
+	// first, check we have not already started a registry (for example, for a different k3d cluster)
+	// all the k3d clusters should share the same private registry, so if we already have a registry just connect
+	// it to the network of this cluster.
+	cid, err := getRegistryContainer()
+	if err != nil {
+		return "", err
+	}
+
+	if cid != "" {
+		// TODO: we should check given-registry-name == existing-registry-name
+		log.Printf("Registry already present: connecting it to the %s network...\n", netName)
+		if err := connectRegistryToNetwork(cid, netName, []string{spec.RegistryName}); err != nil {
+			return "", err
+		}
+		return cid, nil
+	}
+
+	log.Printf("Creating registry as %q...\n", spec.RegistryName)
+
+	containerLabels := make(map[string]string)
+	containerLabels["app"] = "k3d"
+	containerLabels["component"] = "registry"
+	containerLabels["created"] = time.Now().Format("2006-01-02 15:04:05")
+	containerLabels["hostname"] = spec.RegistryName
+
+	registryPortSpec := fmt.Sprintf("0.0.0.0:%d:%d/tcp", spec.RegistryPort, spec.RegistryPort)
+	registryPublishedPorts, err := CreatePublishedPorts([]string{registryPortSpec})
+	if err != nil {
+		log.Fatalf("Error: failed to parse port specs %+v \n%+v", registryPortSpec, err)
+	}
+
+	hostConfig := &container.HostConfig{
+		PortBindings: registryPublishedPorts.PortBindings,
+		Privileged:   true,
+		Init:         &[]bool{true}[0],
+	}
+
+	if spec.AutoRestart {
+		hostConfig.RestartPolicy.Name = "unless-stopped"
+	}
+
+	spec.Volumes = &Volumes{} // we do not need in the registry any of the volumes used by the other containers
+
+	// connect the registry to this k3d network
+	networkingConfig := &network.NetworkingConfig{
+		EndpointsConfig: map[string]*network.EndpointSettings{
+			netName: {
+				Aliases: []string{spec.RegistryName},
+			},
+		},
+	}
+
+	config := &container.Config{
+		Hostname:     spec.RegistryName,
+		Image:        defaultRegistryImage,
+		ExposedPorts: registryPublishedPorts.ExposedPorts,
+		Labels:       containerLabels,
+	}
+
+	id, err := createContainer(config, hostConfig, networkingConfig, defaultRegistryContainerName)
+	if err != nil {
+		return "", fmt.Errorf(" Couldn't create registry container %s\n%w", defaultRegistryContainerName, err)
+	}
+
+	// copy the registry configuration
+	if err := writeRegistriesConfigInContainer(spec, id); err != nil {
```

Comment:
Why do we need the `registries.yaml` (k3s-specific file) in the registry container? :thinking: 
If I'm not wrong, we only need it to make the k3d/k3s nodes aware of the private registry, so we have to put this file into every k3d node container, but not into the registry container :+1:

Reply:
Aghh... good catch. Late night programming...  :sweat:

Reply:
I know that problem :D